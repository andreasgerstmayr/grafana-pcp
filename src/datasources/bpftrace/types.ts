import { Optional } from 'utility-types';
import { MinimalPmapiQuery, PmapiOptions, PmapiQuery } from '../../datasources/lib/pmapi/types';
import { TargetFormat } from '../../datasources/lib/types';
import { Script } from './script';

export interface BPFtraceOptions extends PmapiOptions {}

export interface BPFtraceQuery extends MinimalPmapiQuery {}

export const defaultBPFtraceQuery: BPFtraceQuery & Optional<PmapiQuery, 'url' | 'hostspec'> = {
    refId: 'A',
    expr: '',
    format: TargetFormat.TimeSeries,
    options: {
        rateConversation: true,
        timeUtilizationConversation: true,
    },
};

export interface BPFtraceTargetData {
    script: Script;
}
