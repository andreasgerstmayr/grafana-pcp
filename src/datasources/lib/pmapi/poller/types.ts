import { Context, Indom, InstanceValue, Metadata } from 'common/services/pmapi/types';
import { Labels } from 'common/types/pcp';
import { Dict, RequiredField } from 'common/types/utils';
import { Target, TemplatedPmapiQuery } from '../types';

export interface Metric {
    metadata: Metadata;
    instanceDomain?: {
        instances: Dict<string, Indom>;
        labels: Labels;
    };
    values: InstanceValuesSnapshot[];
}

export interface InstanceValuesSnapshot {
    timestampMs: number;
    values: InstanceValue[];
}

export enum EndpointState {
    /** new entered endpoint, no context available */
    PENDING,
    /** context available */
    CONNECTED,
}

/**
 * single endpoint, identified by url and hostspec
 * each url/hostspec has a different context
 * each url/hostspec can have different metrics (and values)
 */
export interface Endpoint {
    state: EndpointState;
    url: string;
    hostspec: string;
    metrics: Metric[];
    targets: Target[];
    additionalMetricsToPoll: Array<{ name: string; callback: (values: InstanceValue[]) => void }>;
    errors: any[];

    /** context, will be created at next poll */
    context?: Context;
    /** backfilling with redis */
    hasRedis?: boolean;
}

export type EndpointWithCtx = RequiredField<Endpoint, 'context'>;

export interface QueryResult {
    endpoint: EndpointWithCtx;
    query: TemplatedPmapiQuery;
    metrics: Metric[];
}
