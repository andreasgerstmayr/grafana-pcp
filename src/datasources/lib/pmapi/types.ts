import { DataQuery, DataSourceJsonData } from '@grafana/data';
import { RequiredField } from 'common/types/utils';
import { TargetFormat } from '../types';

export interface PmapiOptions extends DataSourceJsonData {
    hostspec?: string;
    retentionTime?: string;
}

export interface PmapiDefaultOptions {
    hostspec: string;
    retentionTime: string;
}

export interface PmapiQuery extends DataQuery {
    expr: string;
    format: TargetFormat;
    legendFormat?: string;

    url?: string;
    hostspec?: string;
}

export type TemplatedPmapiQuery = RequiredField<PmapiQuery, 'url' | 'hostspec'>;

export enum TargetState {
    /** newly entered target or target with error (trying again) */
    PENDING,
    /** metrics exists and metadata available */
    METRICS_AVAILABLE,
    /** fatal error, will not try again */
    ERROR,
}

/**
 * Represents a target of a Grafana panel, which will be polled in the background
 * extends the Query (as provided by Grafana) with additional information:
 * - vector+bpftrace: errors occured while polling in the background, lastActive, ...
 * - vector: isDerivedMetric
 * - bpftrace: script
 * is persisted during multiple queries (key = targetId)
 */
export interface Target<T = any> {
    targetId: string;
    state: TargetState;
    query: TemplatedPmapiQuery;
    /** valid PCP metric names (can be a derived metric, e.g. derived_xxx) */
    metricNames: string[];
    errors: any[];
    lastActiveMs: number;
    custom?: T;
}
