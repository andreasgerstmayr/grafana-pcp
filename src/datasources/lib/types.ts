export enum TargetFormat {
    TimeSeries = 'time_series',
    Heatmap = 'heatmap',
    /** vector only */
    MetricsTable = 'metrics_table',
    /** bpftrace only */
    CsvTable = 'csv_table',
    /** bpftrace only */
    FlameGraph = 'flamegraph',
}
