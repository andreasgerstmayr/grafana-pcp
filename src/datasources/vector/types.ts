import { PmapiOptions, PmapiQuery, TargetFormat } from 'datasources/lib/pmapi/types';

export interface VectorOptions extends PmapiOptions {}

export interface VectorQuery extends PmapiQuery {}

export const defaultVectorQuery: Partial<VectorQuery> = {
    expr: '',
    format: TargetFormat.TimeSeries,
};

export interface VectorTargetData {
    isDerivedMetric: boolean;
}
