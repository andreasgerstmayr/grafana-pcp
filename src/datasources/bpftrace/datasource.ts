import { DataSourceInstanceSettings } from '@grafana/data';
import { keyBy } from 'lodash';
import { BPFtraceQuery, BPFtraceOptions, BPFtraceTargetData } from './types';
import { ScriptManager } from './script_manager';
import { Status, Script } from './script';

import { DatasourceBase } from 'datasources/lib/pmapi/datasource_base';
import { Poller } from 'datasources/lib/pmapi/poller/poller';
import { PmapiQuery, Target, TargetState } from 'datasources/lib/pmapi/types';
import { Endpoint } from 'datasources/lib/pmapi/poller/types';
import { getLogger } from 'common/utils';
import { Config } from './config';
const log = getLogger('datasource');

export class DataSource extends DatasourceBase<BPFtraceQuery, BPFtraceOptions> {
    poller: Poller;
    scriptManager: ScriptManager;

    constructor(readonly instanceSettings: DataSourceInstanceSettings<BPFtraceOptions>) {
        super(instanceSettings, Config.defaults, Config.apiTimeoutMs);
        log.debug('initializate bpftrace datasource');
        this.poller = new Poller(this.pmApiService, this.pmSeriesApiService, {
            retentionTimeMs: this.retentionTimeMs,
            refreshIntervalMs: this.getDashboardRefreshInterval() ?? 1000,
            gracePeriodMs: Config.gracePeriodMs,
            hooks: {
                queryHasChanged: this.queryHasChanged,
                registerEndpoint: this.registerEndpoint.bind(this),
                registerTarget: this.registerTarget.bind(this),
                deregisterTarget: this.deregisterTarget.bind(this),
            },
        });
        this.scriptManager = new ScriptManager(this.pmApiService);

        document.addEventListener('visibilitychange', () => {
            this.poller.setPageVisibility(!document.hidden);
        });
    }

    queryHasChanged(prevQuery: PmapiQuery, newQuery: PmapiQuery) {
        return newQuery.expr !== prevQuery.expr || newQuery.format !== prevQuery.format;
    }

    async registerEndpoint(endpoint: Endpoint) {
        endpoint.additionalMetricsToPoll.push({
            name: 'bpftrace.info.scripts_json',
            callback: values => {
                const scriptsList = JSON.parse(values[0].value as string) as Script[];
                const scriptsById = keyBy(scriptsList, 'script_id');

                for (const target of endpoint.targets as Array<Target<BPFtraceTargetData>>) {
                    const scriptId = target.custom?.script?.script_id;
                    if (!scriptId) {
                        return;
                    }

                    const script = scriptsById[scriptId];
                    if (script?.state.status === Status.Error) {
                        target.errors.push(new Error(`BPFtrace error:\n\n${script.state.error}`));
                        target.state = TargetState.ERROR;
                    }
                }
            },
        });
    }

    async registerTarget(target: Target<BPFtraceTargetData>) {
        const script = await this.scriptManager.register(target.query.url, target.query.hostspec, target.query.expr);
        if (script.state.status === Status.Error) {
            throw new Error(`BPFtrace error:\n\n${script.state.error}`);
        }
        target.custom = { script };
        return this.scriptManager.getMetrics(target.custom.script, target.query.format);
    }

    deregisterTarget(target: Target<BPFtraceTargetData>) {
        if (target.custom?.script) {
            // eslint-disable-next-line @typescript-eslint/no-floating-promises
            this.scriptManager.deregister(target.query.url, target.query.hostspec, target.custom.script);
        }
    }
}
