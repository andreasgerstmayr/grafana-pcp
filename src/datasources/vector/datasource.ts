import { DataSourceInstanceSettings } from '@grafana/data';
import md5 from 'blueimp-md5';
import { VectorQuery, VectorOptions, VectorTargetData } from './types';
import { DataSourceBase } from 'datasources/lib/pmapi/datasource_base';
import { Poller } from 'datasources/lib/pmapi/poller/poller';
import { PmapiQuery, Target } from 'datasources/lib/pmapi/types';
import { Endpoint, InstanceValuesSnapshot, Metric } from 'datasources/lib/pmapi/poller/types';
import { SeriesId, SeriesLabelsItemResponse } from 'common/services/pmseries/types';
import { getLogger } from 'common/utils';
import { Config } from './config';
import { Dict } from 'common/types/utils';
import { InstanceId } from 'common/services/pmapi/types';
import { keyBy } from 'lodash';
const log = getLogger('datasource');

export class PCPVectorDataSource extends DataSourceBase<VectorQuery, VectorOptions> {
    poller: Poller;
    derivedMetrics: Map<string, string>;

    constructor(readonly instanceSettings: DataSourceInstanceSettings<VectorOptions>) {
        super(instanceSettings, Config.defaults, Config.apiTimeoutMs);
        log.debug('initializate Vector datasource');
        this.poller = new Poller(this.pmApiService, this.pmSeriesApiService, {
            retentionTimeMs: this.retentionTimeMs,
            refreshIntervalMs: this.getDashboardRefreshInterval() ?? Config.defaultRefreshIntervalMs,
            gracePeriodMs: Config.gracePeriodMs,
            hooks: {
                queryHasChanged: this.queryHasChanged,
                registerTarget: this.registerTarget.bind(this),
                redisBackfill: this.redisBackfill.bind(this),
            },
        });
        this.derivedMetrics = new Map<string, string>();

        document.addEventListener('visibilitychange', () => {
            this.poller.setPageVisibility(!document.hidden);
        });
    }

    queryHasChanged(prevQuery: PmapiQuery, newQuery: PmapiQuery) {
        return newQuery.expr !== prevQuery.expr;
    }

    isDerivedMetric(expr: string) {
        /* From: PCPIntro(1)
         * A node label must begin with an alphabetic character, followed by
         * zero or more characters drawn from the alphabetics, the digits and
         * character ``_'' (underscore).  For alphabetic characters in a node
         * label, upper and lower case are distinguished.
         *
         * By convention, the name of a performance metric is constructed by
         * concatenation of the node labels on a path through the PMNS from the
         * root node to a leaf node, with a ``.'' as a separator.
         *
         * -> Anything that is not a name is considered derived metric expression
         */
        return !/^[a-zA-Z][a-zA-Z0-9_]*(?:\.[a-zA-Z][a-zA-Z0-9_]+)*$/.test(expr);
    }

    derivedMetricName(expr: string): string {
        return `derived_${md5(expr)}`;
    }

    async registerDerivedMetric(target: Target<VectorTargetData>, endpoint: Endpoint): Promise<string[]> {
        const name = this.derivedMetricName(target.query.expr);
        const context = endpoint.context?.context;
        const result = await this.pmApiService.derive(target.query.url!, {
            context,
            expr: target.query.expr,
            name,
        });
        if (result.success) {
            this.derivedMetrics.set(target.query.expr, name);
            return [name];
        }
        return [];
    }

    async registerTarget(target: Target<VectorTargetData>, endpoint: Endpoint): Promise<string[]> {
        target.custom = {
            isDerivedMetric: this.isDerivedMetric(target.query.expr),
        };
        if (target.custom.isDerivedMetric) {
            const key = this.derivedMetrics.get(target.query.expr);
            if (key) {
                return [key];
            }
            return await this.registerDerivedMetric(target, endpoint);
        } else {
            return [target.query.expr];
        }
    }

    async redisBackfill(endpoint: Endpoint, targets: Array<Target<VectorTargetData>>) {
        const metricNames = new Set(targets.flatMap(target => target.metricNames));

        // split into series (metrics) with and without instance domains
        const seriesWithoutIndom: string[] = [];
        const seriesWithIndom: string[] = [];
        // multiple series can point to the same metric
        const seriesToMetric: Dict<SeriesId, Metric> = {};
        for (const metric of endpoint.metrics) {
            if (metricNames.has(metric.metadata.name)) {
                if (metric.metadata.indom) {
                    seriesWithIndom.push(metric.metadata.series);
                } else {
                    seriesWithoutIndom.push(metric.metadata.series);
                }
                seriesToMetric[metric.metadata.series] = metric;
            }
        }

        // get metric values and instances
        const seekStart = this.retentionTimeMs / 1000;
        const [values, instancesResponse] = await Promise.all([
            this.pmSeriesApiService.values({
                series: [...seriesWithoutIndom, ...seriesWithIndom],
                interval: '1s',
                start: `-${seekStart}second`,
                finish: 'now',
            }),
            seriesWithIndom.length > 0
                ? this.pmSeriesApiService.instances({ series: seriesWithIndom })
                : Promise.resolve([]),
        ]);

        // get instance labels
        let instanceLabels: Dict<string, SeriesLabelsItemResponse> = {};
        if (seriesWithIndom.length > 0) {
            const labelsResponse = (await this.pmSeriesApiService.labels({
                series: instancesResponse.map(instance => instance.instance),
            })) as SeriesLabelsItemResponse[];
            instanceLabels = keyBy(labelsResponse, 'series');
        }

        // fill instanceDomain data structure of poller
        const seriesInstanceToInstanceId: Dict<string, InstanceId> = {};
        for (const instance of instancesResponse) {
            seriesInstanceToInstanceId[instance.instance] = instance.id;

            const metric = seriesToMetric[instance.series]!;
            metric.instanceDomain!.instances[instance.id] = {
                name: instance.name,
                instance: instance.id,
                labels: instanceLabels[instance.instance]?.labels || {},
            };
        }

        // make sure values are sorted series < timestamp
        values.sort((a, b) => {
            if (a.series !== b.series) {
                return a.series < b.series ? -1 : a.series > b.series ? 1 : 0;
            }
            return a.timestamp - b.timestamp;
        });

        // backfill metric values into poller datastructure
        for (let i = 0; i < values.length; ) {
            const curSeries = values[i].series;
            while (i < values.length && values[i].series === curSeries) {
                const curSnapshot: InstanceValuesSnapshot = {
                    timestampMs: values[i].timestamp,
                    values: [],
                };

                // if the timestamp significantly changed from the previous read one, consider it a new measurement
                for (
                    ;
                    i < values.length &&
                    values[i].series === curSeries &&
                    Math.abs(values[i].timestamp - curSnapshot.timestampMs) < 0.1;
                    i++
                ) {
                    const instanceId = values[i].instance ? seriesInstanceToInstanceId[values[i].instance!]! : null;
                    curSnapshot.values.push({ instance: instanceId, value: parseFloat(values[i].value) });
                }

                seriesToMetric[curSeries]!.values.push(curSnapshot);
            }
        }
    }
}
