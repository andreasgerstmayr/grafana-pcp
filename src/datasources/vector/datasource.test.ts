import { advanceTo } from 'jest-date-mock';
import { PCPVectorDataSource } from './datasource';
import { VectorTargetData } from './types';
import { Target } from 'datasources/lib/pmapi/types';
import { Endpoint } from 'datasources/lib/pmapi/poller/types';
import { backendSrvMock, mockNextResponses } from 'datasources/lib/specs/mocks/backend_srv';
import { TargetFormat } from 'datasources/lib/types';
import { grafana, pcp, pmapi, pmseries, poller } from 'datasources/lib/specs/fixtures';
import { setGlobalLogLevel } from 'common/utils';

jest.mock('@grafana/runtime', () => ({
    ...jest.requireActual<object>('@grafana/runtime'),
    getBackendSrv: () => backendSrvMock,
    getTemplateSrv: () => ({
        replace: (x: string) => x,
    }),
}));

describe('PCP Vector', () => {
    let datasource: PCPVectorDataSource;

    beforeEach(() => {
        jest.resetAllMocks();
        jest.useFakeTimers();
        advanceTo(20000);
        setGlobalLogLevel('DEBUG');

        const instanceSettings = {
            url: 'http://localhost:1234',
            jsonData: {},
        };
        datasource = new PCPVectorDataSource(instanceSettings as any);
    });

    it('should poll disk.dev.read, perform rate conversation and return the result', async () => {
        const targets = [{ refId: 'A', expr: 'disk.dev.read', format: TargetFormat.TimeSeries }];

        let response = await datasource.query(grafana.dataQueryRequest(targets));
        expect(response).toEqual({ data: [] });

        mockNextResponses([
            pmapi.context(),
            pmseries.ping(false),
            pmapi.metric(['disk.dev.read']),
            pmapi.fetchDiskDevRead(10, 100, 0),
            pmapi.indom('disk.dev.read'),
        ]);
        await datasource.poller.poll();

        mockNextResponses([pmapi.fetchDiskDevRead(11, 200, 0)]);
        await datasource.poller.poll();

        response = await datasource.query(grafana.dataQueryRequest(targets));
        expect(response).toMatchSnapshot();

        expect(backendSrvMock.fetch.mock.calls).toMatchSnapshot();
    });

    // anything thats not a metric name
    it('should be able to detect derived metric formulas', () => {
        const metrics = ['statsd.pmda.received', 'statsd.pmda', 'statsd', 'statsd_pmda', 'statsd_pmda_received'];
        metrics.forEach(metric => expect(datasource.isDerivedMetric(metric)).toBe(false));

        const formulas = [
            'disk.all.blktotal/disk.all.total',
            'disk.all.blktotal+disk.all.total',
            'disk.all.blktotal-disk.all.total',
            'disk.all.blktotal*disk.all.total',
            'network.interface.in.bytes[eth1]',
            'disk.all.blktotal/2',
            'disk.all.blktotal+2',
            'disk.all.blktotal-2',
            'disk.all.blktotal.2',
            'avg(network.interface.speed)',
        ];
        formulas.forEach(formula => expect(datasource.isDerivedMetric(formula)).toBe(true));
    });

    it('should be able to build a derived metric name', () => {
        const formulas = [
            'disk.all.blktotal/disk.all.total',
            'network.interface.in.bytes[eth1]',
            'disk.all.blktotal/2',
            'disk.all.blktotal+2',
        ];
        const names = formulas.map(formula => {
            const name = datasource.derivedMetricName(formula);
            expect(name).toContain('derived_');
            return name;
        });

        // check that all derived metric names are different
        expect(new Set(names).size).toBe(names.length);
    });

    it('should be able to create derived metric and store information about it', async () => {
        const spy = jest.spyOn(datasource.pmApiService, 'derive');
        const expr = 'disk.all.blktotal/2';
        const targetMock: jest.Mocked<Target<VectorTargetData>> = { query: { expr } } as any;
        const endpointMock: jest.Mocked<Endpoint> = { context: { context: 0 } } as any;
        mockNextResponses([pmapi.derive()]);
        await datasource.registerDerivedMetric(targetMock, endpointMock);

        expect(spy).toBeCalledTimes(1);
        expect(spy.mock.calls[0][1]).toMatchObject({ name: datasource.derivedMetricName(expr) });
        expect(datasource.derivedMetrics.has(expr)).toBe(true);
    });

    it('should request registration of derived metric', async () => {
        const expr = 'disk.all.blktotal/2';
        const metricName = datasource.derivedMetricName(expr);
        const createDerivedSpy = jest.spyOn(datasource.pmApiService, 'derive');
        const registerDeriverMetricSpy = jest.spyOn(datasource, 'registerDerivedMetric');
        const targetMock: jest.Mocked<Target<VectorTargetData>> = { query: { expr } } as any;
        const endpointMock: jest.Mocked<Endpoint> = { context: { context: 0 } } as any;

        mockNextResponses([pmapi.derive()]);
        const resultRegistered = await datasource.registerTarget(targetMock, endpointMock);

        expect(resultRegistered).toEqual([metricName]);
        expect(createDerivedSpy).toBeCalledTimes(1);
        expect(registerDeriverMetricSpy).toBeCalledTimes(1);
        expect(datasource.derivedMetrics.size).toBe(1);
        expect(datasource.derivedMetrics.has(expr)).toBe(true);

        // will skip registering derived metric, since we already did so
        const resultRegistrationSkipped = await datasource.registerTarget(targetMock, endpointMock);
        expect(resultRegistrationSkipped).toEqual([metricName]);
        expect(createDerivedSpy).toBeCalledTimes(1);
        expect(registerDeriverMetricSpy).toBeCalledTimes(1);
        expect(datasource.derivedMetrics.size).toBe(1);
        expect(datasource.derivedMetrics.has(expr)).toBe(true);
    });

    it('redisBackfill hook should request series api for backfilling and populate metric values', async () => {
        const targets = [
            poller.target({ query: { expr: 'disk.dev.read', refId: 'A' } }),
            poller.target({ query: { expr: 'kernel.all.sysfork', refId: 'B' } }),
        ];
        const metrics = [pcp.metrics['disk.dev.read'], pcp.metrics['kernel.all.sysfork']];
        const endpoint = poller.endpoint({ metrics, targets });

        mockNextResponses([
            pmseries.values(['disk.dev.read', 'kernel.all.sysfork']),
            pmseries.instances(['disk.dev.read']),
            pmseries.labels(['disk.dev.read[sda]', 'disk.dev.read[nvme0n1]']),
        ]);
        await datasource.redisBackfill(endpoint, targets);
        expect(endpoint.metrics).toMatchSnapshot();
        expect(backendSrvMock.fetch.mock.calls).toMatchSnapshot();
    });
});
