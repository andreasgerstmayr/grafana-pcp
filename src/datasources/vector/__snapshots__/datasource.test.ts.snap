// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PCP Vector redisBackfill hook should request series api for backfilling and populate metric values 1`] = `
Array [
  Object {
    "instanceDomain": Object {
      "instances": Object {
        "0": Object {
          "instance": 0,
          "labels": Object {
            "agent": "linux",
            "device_type": "block",
            "domainname": "localdomain",
            "groupid": 976,
            "hostname": "dev",
            "indom_name": "per disk",
            "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
            "userid": 978,
          },
          "name": "nvme0n1",
        },
        "1": Object {
          "instance": 1,
          "labels": Object {
            "agent": "linux",
            "device_type": "block",
            "domainname": "localdomain",
            "groupid": 976,
            "hostname": "dev",
            "indom_name": "per disk",
            "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
            "userid": 978,
          },
          "name": "sda",
        },
      },
      "labels": Object {},
    },
    "metadata": Object {
      "indom": "60.1",
      "labels": Object {
        "agent": "linux",
        "device_type": "block",
        "domainname": "localdomain",
        "hostname": "dev",
        "indom_name": "per disk",
        "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
      },
      "name": "disk.dev.read",
      "sem": "counter",
      "series": "f87250c4ea0e5eca8ff2ca3b3044ba1a6c91a3d9",
      "text-help": "Cumulative number of disk read operations since system boot time (subject
to counter wrap).",
      "text-oneline": "per-disk read operations",
      "type": "u64",
      "units": "count",
    },
    "values": Array [
      Object {
        "timestampMs": 1599320691309.872,
        "values": Array [
          Object {
            "instance": 0,
            "value": 1786894,
          },
          Object {
            "instance": 1,
            "value": 0,
          },
        ],
      },
      Object {
        "timestampMs": 1599320691310.872,
        "values": Array [
          Object {
            "instance": 0,
            "value": 1786896,
          },
          Object {
            "instance": 1,
            "value": 0,
          },
        ],
      },
    ],
  },
  Object {
    "metadata": Object {
      "labels": Object {
        "agent": "linux",
        "domainname": "localdomain",
        "hostname": "dev",
        "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
      },
      "name": "kernel.all.sysfork",
      "sem": "counter",
      "series": "73d93ee9efa086923d0c9eabc96f98f2b583b8f2",
      "text-help": "fork rate metric from /proc/stat",
      "text-oneline": "fork rate metric from /proc/stat",
      "type": "u64",
      "units": "count",
    },
    "values": Array [
      Object {
        "timestampMs": 1599320691309.872,
        "values": Array [
          Object {
            "instance": null,
            "value": 100,
          },
        ],
      },
      Object {
        "timestampMs": 1599320691310.872,
        "values": Array [
          Object {
            "instance": null,
            "value": 200,
          },
        ],
      },
    ],
  },
]
`;

exports[`PCP Vector redisBackfill hook should request series api for backfilling and populate metric values 2`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "params": Object {
        "finish": "now",
        "interval": "1s",
        "series": "73d93ee9efa086923d0c9eabc96f98f2b583b8f2,f87250c4ea0e5eca8ff2ca3b3044ba1a6c91a3d9",
        "start": "-1800second",
      },
      "showSuccessAlert": false,
      "url": "http://localhost:1234/series/values",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "params": Object {
        "series": "f87250c4ea0e5eca8ff2ca3b3044ba1a6c91a3d9",
      },
      "showSuccessAlert": false,
      "url": "http://localhost:1234/series/instances",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "params": Object {
        "series": "0aeab8b239522ab0640577ed788cc601fc640266,7f3afb6f41e53792b18e52bcec26fdfa2899fa58",
      },
      "showSuccessAlert": false,
      "url": "http://localhost:1234/series/labels",
    },
  ],
]
`;

exports[`PCP Vector should poll disk.dev.read, perform rate conversation and return the result 1`] = `
Object {
  "data": Array [
    Object {
      "fields": Array [
        Object {
          "config": Object {},
          "labels": undefined,
          "name": "Time",
          "type": "time",
          "values": Array [
            10000,
            11000,
          ],
        },
        Object {
          "config": Object {
            "custom": Object {
              "instance": Object {
                "instance": 0,
                "labels": Object {
                  "domainname": "localdomain",
                  "hostname": "dev",
                  "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
                },
                "name": "nvme0n1",
              },
              "instanceId": 0,
            },
            "displayName": "",
            "unit": undefined,
          },
          "labels": Object {
            "agent": "linux",
            "device_type": "block",
            "domainname": "localdomain",
            "hostname": "dev",
            "indom_name": "per disk",
            "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
          },
          "name": "disk.dev.read[nvme0n1]",
          "type": "number",
          "values": Array [
            null,
            100,
          ],
        },
        Object {
          "config": Object {
            "custom": Object {
              "instance": Object {
                "instance": 1,
                "labels": Object {
                  "domainname": "localdomain",
                  "hostname": "dev",
                  "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
                },
                "name": "sda",
              },
              "instanceId": 1,
            },
            "displayName": "",
            "unit": undefined,
          },
          "labels": Object {
            "agent": "linux",
            "device_type": "block",
            "domainname": "localdomain",
            "hostname": "dev",
            "indom_name": "per disk",
            "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
          },
          "name": "disk.dev.read[sda]",
          "type": "number",
          "values": Array [
            null,
            0,
          ],
        },
      ],
      "meta": Object {
        "custom": Object {
          "context": Object {
            "context": 123,
            "labels": Object {},
          },
          "metric": Object {
            "instanceDomain": Object {
              "instances": Object {
                "0": Object {
                  "instance": 0,
                  "labels": Object {
                    "domainname": "localdomain",
                    "hostname": "dev",
                    "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
                  },
                  "name": "nvme0n1",
                },
                "1": Object {
                  "instance": 1,
                  "labels": Object {
                    "domainname": "localdomain",
                    "hostname": "dev",
                    "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
                  },
                  "name": "sda",
                },
              },
              "labels": Object {
                "device_type": "block",
                "domainname": "localdomain",
                "hostname": "dev",
                "indom_name": "per disk",
                "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
              },
            },
            "metadata": Object {
              "indom": "60.1",
              "labels": Object {
                "agent": "linux",
                "device_type": "block",
                "domainname": "localdomain",
                "hostname": "dev",
                "indom_name": "per disk",
                "machineid": "6dabb302d60b402dabcc13dc4fd0fab8",
              },
              "name": "disk.dev.read",
              "sem": "counter",
              "series": "f87250c4ea0e5eca8ff2ca3b3044ba1a6c91a3d9",
              "text-help": "Cumulative number of disk read operations since system boot time (subject
to counter wrap).",
              "text-oneline": "per-disk read operations",
              "type": "u64",
              "units": "count",
            },
            "values": Array [
              Object {
                "timestampMs": 10000,
                "values": Array [
                  Object {
                    "instance": 0,
                    "value": 100,
                  },
                  Object {
                    "instance": 1,
                    "value": 0,
                  },
                ],
              },
              Object {
                "timestampMs": 11000,
                "values": Array [
                  Object {
                    "instance": 0,
                    "value": 200,
                  },
                  Object {
                    "instance": 1,
                    "value": 0,
                  },
                ],
              },
            ],
          },
          "query": Object {
            "expr": "disk.dev.read",
            "format": "time_series",
            "hostspec": "127.0.0.1",
            "refId": "A",
            "url": "http://localhost:1234",
          },
        },
      },
      "name": undefined,
      "refId": undefined,
    },
  ],
}
`;

exports[`PCP Vector should poll disk.dev.read, perform rate conversation and return the result 2`] = `
Array [
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "params": Object {
        "hostspec": "127.0.0.1",
        "polltimeout": 11,
      },
      "showSuccessAlert": false,
      "url": "http://localhost:1234/pmapi/context",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "showSuccessAlert": false,
      "url": "http://localhost:1234/series/ping",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "params": Object {
        "context": 123,
        "names": "disk.dev.read",
      },
      "showSuccessAlert": false,
      "url": "http://localhost:1234/pmapi/metric",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "params": Object {
        "context": 123,
        "names": "disk.dev.read",
      },
      "showSuccessAlert": false,
      "url": "http://localhost:1234/pmapi/fetch",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "params": Object {
        "context": 123,
        "name": "disk.dev.read",
      },
      "showSuccessAlert": false,
      "url": "http://localhost:1234/pmapi/indom",
    },
  ],
  Array [
    Object {
      "headers": Object {
        "Content-Type": "application/json",
      },
      "params": Object {
        "context": 123,
        "names": "disk.dev.read",
      },
      "showSuccessAlert": false,
      "url": "http://localhost:1234/pmapi/fetch",
    },
  ],
]
`;
